### PR TITLE
Loading products from mock backend

### PR DESCRIPTION
**Pull request by gadgetfather**
## Files to Check

- [ ] ProductPage.jsx
- [ ] product-context.js
- [ ] ProductGrid.jsx
- [ ] index.js
- [ ] Loader.css
- [ ] Loader.jsx

## Preview
Animation

### CODE REVIEW

File: src/context/product-context.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+import axios from "axios";
+import {
+  useReducer,
+  createContext,
+  useContext,
+  useState,
+  useEffect,
+} from "react";
+
+const productContext = createContext();
```
Comment (AtulPant2704) [inline src/context/product-context.js]: In productContext, "P" should be capital.

**Review by mohammed-sas — APPROVED** [review]
1) liked your loader design
2) product card design looks great
3) neatly structured folder

File: src/components/ProductGrid/ProductGrid.jsx [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+import { useProduct } from "../../context/product-context";
 export function ProductGrid() {
+  const state = useProduct();
+  const { products, loader } = state;
+  console.log(products);
   return (
     <div className="product-section">
-      {[1, 2, 3, 5, 1, 1, 1, 1].map((item, indx) => (
-        <ProductCard key={indx} />
+      {console.log("loader", loader)}
```
Comment (AtulPant2704) [inline src/components/ProductGrid/ProductGrid.jsx]: You can remove both of these console.log, as you are already done with testing and now you have deployed it.

**Review by AtulPant2704 — COMMENTED** [review]
I saw that you have used useContext for loading Products and applied that context to the whole app, but because of that your products are getting fetched while the user is in the homepage itself. While I think it is better if the products gets fetched only from the productListing page. Could you please tell me why have you used context ? You can achieve same thing using useEffect for loading products and useState for changing loader in productListing page itself.

File: src/pages/ProductPage/ProductPage.jsx [hunk 3]
```diff
@@ -1,12 +1,25 @@
...
 export function ProductPage() {
+  const [loader, setLoader] = useState(true);
+  const [products, setProducts] = useState([]);
+
+  async function getProducts() {
+    setLoader(true);
+    const { data } = await axios.get("./api/products");
+    setLoader(false);
+    setProducts(data.products);
+  }
```
Comment (AtulPant2704) [inline src/pages/ProductPage/ProductPage.jsx]: You should use try catch for the server call.

Comment (gadgetfather) [reply src/pages/ProductPage/ProductPage.jsx]: Yes I will add that

File: src/pages/ProductPage/ProductPage.jsx [hunk 4]
```diff
@@ -8,10 +8,14 @@ export function ProductPage() {
...
-    setLoader(true);
-    const { data } = await axios.get("./api/products");
-    setLoader(false);
-    setProducts(data.products);
+    try {
+      setLoader(true);
+      const { data } = await axios.get("./api/products");
+      setLoader(false);
+      setProducts(data.products);
+    } catch (error) {
```
Comment (AtulPant2704) [inline src/pages/ProductPage/ProductPage.jsx:16]: Later on you can display here error as well, for now you can just console.log the error.

**Review by AtulPant2704 — APPROVED** [review]
PR looks good to merge now.

**Pull request merged**